### PR TITLE
Add default values

### LINKED ISSUE #6: Add "release" feature
The Raven Sentry package allows developers to display and send the current GIT sha to Sentry itself for tracking of the current "release". This should be added ASAP.

### PR DESCRIPTION
**Pull request by Firesphere**
- Set the release to the hash if possible
- Add the message as an extra if possible

Partially fixes #6 and adds the last commit message as an extra bit of information.

### CODE REVIEW

**Comment by Firesphere** [comment]
The commands are prefixed with an `@`, to prevent it from throwing an error on it's own if the file doesn't exist or it doesn't have read access.
Or, possibly, file_get_contents etc. does not have execution permissions.

**Comment by phptek** [comment]
You could probably abstract the GIT file logic into a private method on `SentryLogger` e.g. `getGitfileContents()` and then in one fell-swoop include the extra data:

Incomplete pseudo-code:

```
    public function defaultExtra() : array
    {
        $ref = $this->getGitfileContents($head[1]);
        $this->client->getOptions()->setRelease($ref);
        
        return [
            'Peak-Memory'   => $this->getPeakMemory(),
            'Commit Message' => $ref = $this->getGitfileContents('COMMIT_EDITMSG'),
        ];
    }
```

**Comment by Firesphere** [comment]
Good point :+1:

**Comment by Firesphere** [comment]
I've added some coverages etc. as well, so you could set up CodeClimate/CodeCov/Scrutinizer-CI (Free for OSS projects)

**Comment by Firesphere** [comment]
I did some code updates

**Comment by Firesphere** [comment]
Ping @phptek

**Comment by phptek** [comment]
Yeah, I've seen it. Been sick and busy recently.

**Comment by Firesphere** [comment]
Ow, apologies! Get well soon!

File: src/Log/SentryLogger.php [hunk 1]
```diff
@@ -220,81 +221,81 @@ public function getReqMethod() : string
...
-		}
+        if ($headerOverrideIP) {
+            return $proxy->getIPFromHeaderValue($headerOverrideIP);
+        }
 
         if (isset($_SERVER['REMOTE_ADDR'])) {
-			return $_SERVER['REMOTE_ADDR'];
-		}
+            return $_SERVER['REMOTE_ADDR'];
+        }
```
Comment (phptek) [inline src/Log/SentryLogger.php]: `return $_SERVER['REMOTE_ADDR'] ?? 'N/A'` ?

File: src/Log/SentryLogger.php [hunk 2]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
         ]);
     }
 
+    /**
+     * Get the information about the current release, if possible
+     * @return array
```
Comment (phptek) [inline src/Log/SentryLogger.php]: I'm using casted returns everywhere these days, so an appended `: array` would be grand :-)

File: src/Log/SentryLogger.php [hunk 3]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+    }
+
+    /**
+     * Get the HEAD tag or the release
+     *
+     * @return string
+     */
+    public static function getGitHead()
+    {
+        $head = file_exists(Director::baseFolder() . '/.git/HEAD');
```
Comment (phptek) [inline src/Log/SentryLogger.php]: I pulled your branch in order to test this, but the logic there looks different (Misplaced `/`) resulting in file not found. Not sure what that's about.

File: src/Log/SentryLogger.php [hunk 4]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+        if ($head) {
+            $head = explode(':', file_get_contents(Director::baseFolder() . '/.git/HEAD'));
```
Comment (phptek) [inline src/Log/SentryLogger.php]: You've already made a F/S call using `file_get_contents()` to build your `$head` var, why not re-use it and avoid another?

```
if ($head = file_exists(Director::baseFolder() . '/.git/HEAD')) {
    // Do stuff with $head
    ...
}
```

File: src/Log/SentryLogger.php [hunk 5]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+
+        return '';
+    }
+
+    /**
+     * Check if we are on a tagged release and if so, use that to set the client release
+     * @param $ref
+     * @return mixed
+     */
+    protected static function getGitTag($ref)
```
Comment (phptek) [inline src/Log/SentryLogger.php]: As above, type-hinting and casted returns please

File: src/Log/SentryLogger.php [hunk 6]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+        $data = file_get_contents(Director::baseFolder() . './git/packed-refs');
+        if ($data) {
+            $data = explode('\n', $data);
+            foreach ($data as $line) {
+                if (strpos($line, '#') !== false) {
+                    continue;
+                }
+                $tags = explode(' ', $line);
+                if ($tags[0] === $ref) {
+                    list(,,$tag) = explode('/', $ref[1]);
```
Comment (phptek) [inline src/Log/SentryLogger.php]: TIL ^^

File: src/Log/SentryLogger.php [hunk 7]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+        return $ref;
+    }
+
+    /**
+     * @param array $return
+     * @return array
+     */
+    protected function getGitCommitMessage(array $return): array
+    {
+        $msgFile = file_exists(Director::baseFolder() . '/.git/COMMIT_EDITMSG');
```
Comment (phptek) [inline src/Log/SentryLogger.php]: Not a massive deal, but if logic needs to make so many F/S calls to achieve its aims, why not do it once for each required file, and cache the result as an array of your selected files, to the current scope?

File: src/Log/SentryLogger.php [hunk 8]
```diff
@@ -332,4 +333,80 @@ public static function backtrace(array $record) : array
...
+    /**
+     * Check if we are on a tagged release and if so, use that to set the client release
+     * @param $ref
+     * @return mixed
+     */
+    protected static function getGitTag($ref)
+    {
+        $data = file_get_contents(Director::baseFolder() . './git/packed-refs');
+        if ($data) {
+            $data = explode('\n', $data);
```
Comment (phptek) [inline src/Log/SentryLogger.php]: This doesn't do what you think it does. You need either `"\n"` or even better `PHP_EOL` to serve Windows envs.

**Comment by phptek** [comment]
This work needs to be rebased against `master` (My fault, for being sick and then just plain tardy). Once you've attended to the comments above, please squash the x3 commitss. Thanks!

**Comment by Firesphere** [comment]
I totally did not have time. I've put it on my to do list for this week, to fix everything up properly!

**Comment by Firesphere** [comment]
Nope, time said no :/

**Comment by phptek** [comment]
Closing this in favour of using `Composer\InstalledVersions::getRootPackage()` called in `SentryLogger::getAppInfo()` in `4.0.0-alpha2` and above. This gives us the project name, branch and commit hash.

**Pull request closed without merging**